File: c_programs_and_containers/build_blocks/knowledge_base/mqtt/kv_read_store/src/kv_store_writer.c
/**
 * kv_store_writer.c - Writes retained key/value messages to an MQTT broker.
 *
 * Translated from Python KVStoreWriter class.
 * Requires: libmosquitto (mosquitto-dev)
 *   apt install libmosquitto-dev
 *   Link with: -lmosquitto -lpthread
 */

#include "kv_store_writer.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <math.h>
#include <mosquitto.h>

/* ── Internal helpers ───────────────────────────────────────────────── */

static void kvw_now(struct timespec *ts)
{
    clock_gettime(CLOCK_REALTIME, ts);
}

/**
 * Build an absolute timespec that is `offset_sec` seconds from now.
 */
static void kvw_deadline(struct timespec *out, double offset_sec)
{
    kvw_now(out);
    long whole = (long)offset_sec;
    long frac  = (long)((offset_sec - (double)whole) * 1e9);
    out->tv_nsec += frac;
    out->tv_sec  += whole + out->tv_nsec / 1000000000L;
    out->tv_nsec  = out->tv_nsec % 1000000000L;
}

/* ── Pending-publish helpers (caller must hold pending_lock) ────────── */

/**
 * Register a new mid to track.  Returns slot index or -1 if full.
 */
static int kvw_pending_add(kvw_store_writer_t *kv, int mid)
{
    if (kv->pending_count >= KVW_MAX_PENDING) return -1;
    int idx = kv->pending_count++;
    kv->pending[idx].mid       = mid;
    kv->pending[idx].completed = false;
    kv->pending[idx].success   = false;
    return idx;
}

/**
 * Find a pending slot by mid. Returns index or -1.
 */
static int kvw_pending_find(kvw_store_writer_t *kv, int mid)
{
    for (int i = 0; i < kv->pending_count; i++) {
        if (kv->pending[i].mid == mid) return i;
    }
    return -1;
}

/**
 * Remove a pending slot by index (swap with last).
 */
static void kvw_pending_remove(kvw_store_writer_t *kv, int idx)
{
    if (idx < 0 || idx >= kv->pending_count) return;
    kv->pending[idx] = kv->pending[kv->pending_count - 1];
    kv->pending_count--;
}

/**
 * Wait for a specific mid to complete.
 * Caller must hold pending_lock on entry; lock is held on return.
 * Returns true if completed before deadline, false on timeout.
 */
static bool kvw_pending_wait(kvw_store_writer_t *kv, int mid,
                             const struct timespec *deadline)
{
    for (;;) {
        int idx = kvw_pending_find(kv, mid);
        if (idx < 0) return false;               /* was removed externally */
        if (kv->pending[idx].completed) return true;

        int rc = pthread_cond_timedwait(&kv->pending_cond,
                                        &kv->pending_lock, deadline);
        if (rc == ETIMEDOUT) return false;
    }
}

/* ── Mosquitto callbacks ────────────────────────────────────────────── */

static void on_connect_cb(struct mosquitto *mosq, void *obj, int reason_code)
{
    (void)mosq;
    kvw_store_writer_t *kv = (kvw_store_writer_t *)obj;

    pthread_mutex_lock(&kv->state_lock);
    kv->connected    = (reason_code == 0);
    kv->connect_done = true;
    pthread_cond_signal(&kv->connect_cond);
    pthread_mutex_unlock(&kv->state_lock);

    if (kv->connected) {
        printf("[Connected] Successfully connected to %s:%d\n",
               kv->config.host, kv->config.port);
    } else {
        printf("[Connection Failed] reason_code=%d\n", reason_code);
    }
}

static void on_disconnect_cb(struct mosquitto *mosq, void *obj, int reason_code)
{
    (void)mosq;
    kvw_store_writer_t *kv = (kvw_store_writer_t *)obj;

    pthread_mutex_lock(&kv->state_lock);
    kv->connected = false;
    pthread_mutex_unlock(&kv->state_lock);

    printf("[Disconnected] reason_code=%d\n", reason_code);
}

static void on_publish_cb(struct mosquitto *mosq, void *obj, int mid)
{
    (void)mosq;
    kvw_store_writer_t *kv = (kvw_store_writer_t *)obj;

    pthread_mutex_lock(&kv->pending_lock);
    int idx = kvw_pending_find(kv, mid);
    if (idx >= 0) {
        kv->pending[idx].completed = true;
        kv->pending[idx].success   = true;
    }
    pthread_cond_broadcast(&kv->pending_cond);
    pthread_mutex_unlock(&kv->pending_lock);
}

/* ── Public API ─────────────────────────────────────────────────────── */

void kvw_config_init(kvw_config_t *cfg)
{
    memset(cfg, 0, sizeof(*cfg));
    strncpy(cfg->host, "localhost", sizeof(cfg->host) - 1);
    cfg->port            = 1883;
    strncpy(cfg->client_id, "kv-writer", sizeof(cfg->client_id) - 1);
    cfg->keepalive       = 60;
    cfg->use_mqttv5      = false;
    cfg->has_credentials = false;
    cfg->clean_session   = true;
}

int kvw_init(kvw_store_writer_t *kv, const kvw_config_t *cfg)
{
    memset(kv, 0, sizeof(*kv));
    kv->config = *cfg;

    kv->mosq = mosquitto_new(cfg->client_id, cfg->clean_session, kv);
    if (!kv->mosq) {
        fprintf(stderr, "[Error] mosquitto_new failed: %s\n", strerror(errno));
        return -1;
    }

    if (cfg->has_credentials) {
        mosquitto_username_pw_set(kv->mosq, cfg->username, cfg->password);
    }

    if (cfg->use_mqttv5) {
        int ver = MQTT_PROTOCOL_V5;
        mosquitto_int_option(kv->mosq, MOSQ_OPT_PROTOCOL_VERSION, ver);
    }

    mosquitto_connect_callback_set(kv->mosq, on_connect_cb);
    mosquitto_disconnect_callback_set(kv->mosq, on_disconnect_cb);
    mosquitto_publish_callback_set(kv->mosq, on_publish_cb);

    pthread_mutex_init(&kv->pending_lock, NULL);
    pthread_cond_init(&kv->pending_cond, NULL);
    pthread_mutex_init(&kv->state_lock, NULL);
    pthread_cond_init(&kv->connect_cond, NULL);

    return 0;
}

bool kvw_connect(kvw_store_writer_t *kv, double timeout_sec)
{
    pthread_mutex_lock(&kv->state_lock);
    kv->connect_done = false;
    kv->running      = true;
    pthread_mutex_unlock(&kv->state_lock);

    int rc = mosquitto_connect(kv->mosq, kv->config.host,
                               kv->config.port, kv->config.keepalive);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "[Error] Connection failed: %s\n", mosquitto_strerror(rc));
        kv->running = false;
        return false;
    }

    rc = mosquitto_loop_start(kv->mosq);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "[Error] loop_start failed: %s\n", mosquitto_strerror(rc));
        kv->running = false;
        return false;
    }

    /* Wait for CONNACK */
    struct timespec deadline;
    kvw_deadline(&deadline, timeout_sec);

    pthread_mutex_lock(&kv->state_lock);
    while (!kv->connect_done) {
        int wait_rc = pthread_cond_timedwait(&kv->connect_cond,
                                             &kv->state_lock, &deadline);
        if (wait_rc == ETIMEDOUT) break;
    }
    bool ok = kv->connected;
    pthread_mutex_unlock(&kv->state_lock);

    if (!ok) {
        fprintf(stderr, "[Error] Connection timeout or refused (%s:%d)\n",
                kv->config.host, kv->config.port);
        kv->running = false;
        mosquitto_loop_stop(kv->mosq, true);
        return false;
    }

    return true;
}

void kvw_disconnect(kvw_store_writer_t *kv)
{
    kv->running = false;
    mosquitto_disconnect(kv->mosq);
    mosquitto_loop_stop(kv->mosq, false);

    pthread_mutex_lock(&kv->state_lock);
    kv->connected = false;
    pthread_mutex_unlock(&kv->state_lock);
}

void kvw_destroy(kvw_store_writer_t *kv)
{
    if (kv->running || kv->connected) {
        kvw_disconnect(kv);
    }
    if (kv->mosq) {
        mosquitto_destroy(kv->mosq);
        kv->mosq = NULL;
    }
    pthread_mutex_destroy(&kv->pending_lock);
    pthread_cond_destroy(&kv->pending_cond);
    pthread_mutex_destroy(&kv->state_lock);
    pthread_cond_destroy(&kv->connect_cond);
}

bool kvw_write_single(kvw_store_writer_t *kv,
                      const char *topic,
                      const char *value,
                      int qos,
                      bool retain,
                      double timeout_sec)
{
    if (!kv->connected) {
        fprintf(stderr, "[Error] Not connected. Call kvw_connect() first.\n");
        return false;
    }

    int mid = 0;
    int payloadlen = value ? (int)strlen(value) : 0;
    int rc = mosquitto_publish(kv->mosq, &mid, topic,
                               payloadlen, value, qos, retain);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "[Error] Failed to queue message for %s: %s\n",
                topic, mosquitto_strerror(rc));
        return false;
    }

    /* Register mid and wait for on_publish */
    struct timespec deadline;
    kvw_deadline(&deadline, timeout_sec);

    pthread_mutex_lock(&kv->pending_lock);
    int idx = kvw_pending_add(kv, mid);
    if (idx < 0) {
        pthread_mutex_unlock(&kv->pending_lock);
        fprintf(stderr, "[Error] Too many pending publishes\n");
        return false;
    }

    bool completed = kvw_pending_wait(kv, mid, &deadline);
    bool success   = false;

    if (completed) {
        idx = kvw_pending_find(kv, mid);
        if (idx >= 0) {
            success = kv->pending[idx].success;
            kvw_pending_remove(kv, idx);
        }
    } else {
        fprintf(stderr, "[Timeout] Publish timeout for %s\n", topic);
        idx = kvw_pending_find(kv, mid);
        if (idx >= 0) kvw_pending_remove(kv, idx);
    }
    pthread_mutex_unlock(&kv->pending_lock);

    if (success) {
        /* Truncated preview */
        char preview[54];
        if (value) {
            int plen = (int)strlen(value);
            int clen = plen < 50 ? plen : 50;
            memcpy(preview, value, clen);
            preview[clen] = '\0';
            printf("[Written] %s => %s%s\n", topic, preview,
                   plen > 50 ? "..." : "");
        } else {
            printf("[Written] %s => (null)\n", topic);
        }
    }

    return success;
}

int kvw_write_batch(kvw_store_writer_t *kv,
                    int count,
                    const char *topics[],
                    const char *values[],
                    int qos,
                    bool retain,
                    double timeout_sec,
                    const char *failed_topics[])
{
    if (!kv->connected) {
        fprintf(stderr, "[Error] Not connected. Call kvw_connect() first.\n");
        return 0;
    }
    if (count <= 0) return 0;
    if (count > KVW_MAX_BATCH) count = KVW_MAX_BATCH;

    struct timespec deadline;
    kvw_deadline(&deadline, timeout_sec);

    /* Phase 1: publish all messages and register mids */
    int mids[KVW_MAX_BATCH];
    bool queued[KVW_MAX_BATCH];
    int fail_idx = 0;

    pthread_mutex_lock(&kv->pending_lock);
    for (int i = 0; i < count; i++) {
        mids[i]   = 0;
        queued[i] = false;

        const char *val = values[i] ? values[i] : "";
        int rc = mosquitto_publish(kv->mosq, &mids[i], topics[i],
                                   (int)strlen(val), val, qos, retain);
        if (rc != MOSQ_ERR_SUCCESS) {
            fprintf(stderr, "[Error] Failed to queue %s: %s\n",
                    topics[i], mosquitto_strerror(rc));
            if (failed_topics) failed_topics[fail_idx++] = topics[i];
            continue;
        }

        if (kvw_pending_add(kv, mids[i]) < 0) {
            fprintf(stderr, "[Error] Too many pending publishes\n");
            if (failed_topics) failed_topics[fail_idx++] = topics[i];
            continue;
        }
        queued[i] = true;
    }
    pthread_mutex_unlock(&kv->pending_lock);

    /* Phase 2: wait for each queued message */
    int success_count = 0;

    pthread_mutex_lock(&kv->pending_lock);
    for (int i = 0; i < count; i++) {
        if (!queued[i]) continue;

        bool completed = kvw_pending_wait(kv, mids[i], &deadline);
        int idx = kvw_pending_find(kv, mids[i]);

        if (completed && idx >= 0 && kv->pending[idx].success) {
            success_count++;
            printf("[Batch Written] %s\n", topics[i]);
        } else {
            if (!completed) {
                fprintf(stderr, "[Batch Timeout] %s\n", topics[i]);
            }
            if (failed_topics) failed_topics[fail_idx++] = topics[i];
        }

        if (idx >= 0) kvw_pending_remove(kv, idx);
    }
    pthread_mutex_unlock(&kv->pending_lock);

    return success_count;
}

bool kvw_delete_single(kvw_store_writer_t *kv,
                       const char *topic,
                       double timeout_sec)
{
    if (!kv->connected) {
        fprintf(stderr, "[Error] Not connected. Call kvw_connect() first.\n");
        return false;
    }

    /* Publish empty retained message to delete the key */
    int mid = 0;
    int rc = mosquitto_publish(kv->mosq, &mid, topic, 0, "", 1, true);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "[Error] Failed to delete %s: %s\n",
                topic, mosquitto_strerror(rc));
        return false;
    }

    struct timespec deadline;
    kvw_deadline(&deadline, timeout_sec);

    pthread_mutex_lock(&kv->pending_lock);
    int idx = kvw_pending_add(kv, mid);
    if (idx < 0) {
        pthread_mutex_unlock(&kv->pending_lock);
        fprintf(stderr, "[Error] Too many pending publishes\n");
        return false;
    }

    bool completed = kvw_pending_wait(kv, mid, &deadline);
    bool success   = false;

    if (completed) {
        idx = kvw_pending_find(kv, mid);
        if (idx >= 0) {
            success = kv->pending[idx].success;
            kvw_pending_remove(kv, idx);
        }
    } else {
        fprintf(stderr, "[Timeout] Delete timeout for %s\n", topic);
        idx = kvw_pending_find(kv, mid);
        if (idx >= 0) kvw_pending_remove(kv, idx);
    }
    pthread_mutex_unlock(&kv->pending_lock);

    if (success) {
        printf("[Deleted] %s\n", topic);
    }

    return success;
}

int kvw_delete_batch(kvw_store_writer_t *kv,
                     int count,
                     const char *topics[],
                     double timeout_sec,
                     const char *failed_topics[])
{
    if (!kv->connected) {
        fprintf(stderr, "[Error] Not connected. Call kvw_connect() first.\n");
        return 0;
    }
    if (count <= 0) return 0;
    if (count > KVW_MAX_BATCH) count = KVW_MAX_BATCH;

    struct timespec deadline;
    kvw_deadline(&deadline, timeout_sec);

    /* Phase 1: publish all empty retained messages */
    int mids[KVW_MAX_BATCH];
    bool queued[KVW_MAX_BATCH];
    int fail_idx = 0;

    pthread_mutex_lock(&kv->pending_lock);
    for (int i = 0; i < count; i++) {
        mids[i]   = 0;
        queued[i] = false;

        int rc = mosquitto_publish(kv->mosq, &mids[i], topics[i],
                                   0, "", 1, true);
        if (rc != MOSQ_ERR_SUCCESS) {
            fprintf(stderr, "[Error] Failed to queue delete for %s: %s\n",
                    topics[i], mosquitto_strerror(rc));
            if (failed_topics) failed_topics[fail_idx++] = topics[i];
            continue;
        }

        if (kvw_pending_add(kv, mids[i]) < 0) {
            fprintf(stderr, "[Error] Too many pending publishes\n");
            if (failed_topics) failed_topics[fail_idx++] = topics[i];
            continue;
        }
        queued[i] = true;
    }
    pthread_mutex_unlock(&kv->pending_lock);

    /* Phase 2: wait for each */
    int success_count = 0;

    pthread_mutex_lock(&kv->pending_lock);
    for (int i = 0; i < count; i++) {
        if (!queued[i]) continue;

        bool completed = kvw_pending_wait(kv, mids[i], &deadline);
        int idx = kvw_pending_find(kv, mids[i]);

        if (completed && idx >= 0 && kv->pending[idx].success) {
            success_count++;
            printf("[Batch Deleted] %s\n", topics[i]);
        } else {
            if (!completed) {
                fprintf(stderr, "[Delete Timeout] %s\n", topics[i]);
            }
            if (failed_topics) failed_topics[fail_idx++] = topics[i];
        }

        if (idx >= 0) kvw_pending_remove(kv, idx);
    }
    pthread_mutex_unlock(&kv->pending_lock);

    return success_count;
}

bool kvw_update_single(kvw_store_writer_t *kv,
                       const char *topic,
                       const char *value,
                       int qos,
                       double timeout_sec)
{
    return kvw_write_single(kv, topic, value, qos, true, timeout_sec);
}

bool kvw_is_connected(kvw_store_writer_t *kv)
{
    pthread_mutex_lock(&kv->state_lock);
    bool c = kv->connected;
    pthread_mutex_unlock(&kv->state_lock);
    return c;
}

File: zig_programs_and_containers/building_blocks/knowledge_base/mqtt/kv_read_store/src/kv_store_writer.zig
//! kv_store_writer.zig - Writes retained key/value messages to an MQTT broker.
//!
//! Translated from C KVStoreWriter.
//! Requires: libmosquitto (apt install libmosquitto-dev)

const std = @import("std");
const log = std.log.scoped(.kv_writer);

const c = @cImport({
    @cInclude("mosquitto.h");
});

// ──────────────────────────────────────────────────────────────────────
//  Constants
// ──────────────────────────────────────────────────────────────────────

pub const MAX_TOPIC_LEN = 256;
pub const MAX_VALUE_LEN = 4096;
pub const MAX_PENDING = 128;
pub const MAX_BATCH = 64;

// ──────────────────────────────────────────────────────────────────────
//  Types
// ──────────────────────────────────────────────────────────────────────

pub const Config = struct {
    host: [:0]const u8 = "localhost",
    port: u16 = 1883,
    client_id: [:0]const u8 = "kv-writer",
    keepalive: u16 = 60,
    username: ?[:0]const u8 = null,
    password: ?[:0]const u8 = null,
    clean_session: bool = true,
};

pub const Qos = enum(c_int) {
    at_most_once = 0,
    at_least_once = 1,
    exactly_once = 2,
};

const PendingPublish = struct {
    mid: c_int = 0,
    completed: bool = false,
    success: bool = false,
};

pub const Error = error{
    MosquittoNew,
    ConnectFailed,
    LoopStartFailed,
    ConnectTimeout,
    NotConnected,
    PublishFailed,
    TooManyPending,
    PublishTimeout,
};

// ──────────────────────────────────────────────────────────────────────
//  Writer
// ──────────────────────────────────────────────────────────────────────

pub const Writer = struct {
    mosq: *c.struct_mosquitto = undefined,
    cfg: Config,

    // Publish tracking
    pending: [MAX_PENDING]PendingPublish = [_]PendingPublish{.{}} ** MAX_PENDING,
    pending_count: usize = 0,
    pending_mutex: std.Thread.Mutex = .{},
    pending_cond: std.Thread.Condition = .{},

    // Connection state
    connected: bool = false,
    running: bool = false,
    connect_done: bool = false,
    state_mutex: std.Thread.Mutex = .{},
    connect_cond: std.Thread.Condition = .{},

    const Self = @This();

    // ── Pending helpers (caller must hold pending_mutex) ─────────────

    fn pendingAdd(self: *Self, mid: c_int) ?usize {
        if (self.pending_count >= MAX_PENDING) return null;
        const idx = self.pending_count;
        self.pending[idx] = .{ .mid = mid, .completed = false, .success = false };
        self.pending_count += 1;
        return idx;
    }

    fn pendingFind(self: *Self, mid: c_int) ?usize {
        for (0..self.pending_count) |i| {
            if (self.pending[i].mid == mid) return i;
        }
        return null;
    }

    fn pendingRemove(self: *Self, idx: usize) void {
        if (idx >= self.pending_count) return;
        self.pending[idx] = self.pending[self.pending_count - 1];
        self.pending_count -= 1;
    }

    /// Wait for a specific mid to complete. Caller must hold pending_mutex.
    fn pendingWait(self: *Self, mid: c_int, timeout_ns: u64) bool {
        const deadline_ns = @as(u64, @intCast(std.time.nanoTimestamp())) + timeout_ns;
        while (true) {
            const idx = self.pendingFind(mid) orelse return false;
            if (self.pending[idx].completed) return true;

            const now_ns = @as(u64, @intCast(std.time.nanoTimestamp()));
            if (now_ns >= deadline_ns) return false;
            const remaining = deadline_ns - now_ns;

            self.pending_cond.timedWait(&self.pending_mutex, remaining) catch return false;
        }
    }

    // ── Public API ───────────────────────────────────────────────────

    pub fn init(cfg: Config) Error!Self {
        var self = Self{ .cfg = cfg };

        const mosq = c.mosquitto_new(cfg.client_id.ptr, cfg.clean_session, null);
        if (mosq == null) {
            log.err("[Error] mosquitto_new failed", .{});
            return Error.MosquittoNew;
        }
        self.mosq = mosq.?;

        if (cfg.username) |user| {
            const pw: ?[*:0]const u8 = if (cfg.password) |p| p.ptr else null;
            _ = c.mosquitto_username_pw_set(self.mosq, user.ptr, pw);
        }

        c.mosquitto_connect_callback_set(self.mosq, onConnectCb);
        c.mosquitto_disconnect_callback_set(self.mosq, onDisconnectCb);
        c.mosquitto_publish_callback_set(self.mosq, onPublishCb);

        return self;
    }

    pub fn connect(self: *Self, timeout_ms: u64) Error!void {
        c.mosquitto_user_data_set(self.mosq, @ptrCast(self));

        self.state_mutex.lock();
        self.connect_done = false;
        self.running = true;
        self.state_mutex.unlock();

        const rc_conn = c.mosquitto_connect(
            self.mosq,
            self.cfg.host.ptr,
            @intCast(self.cfg.port),
            @intCast(self.cfg.keepalive),
        );
        if (rc_conn != c.MOSQ_ERR_SUCCESS) {
            log.err("[Error] Connection failed: {s}", .{c.mosquitto_strerror(rc_conn)});
            self.running = false;
            return Error.ConnectFailed;
        }

        const rc_loop = c.mosquitto_loop_start(self.mosq);
        if (rc_loop != c.MOSQ_ERR_SUCCESS) {
            log.err("[Error] loop_start failed: {s}", .{c.mosquitto_strerror(rc_loop)});
            self.running = false;
            return Error.LoopStartFailed;
        }

        // Wait for CONNACK
        self.state_mutex.lock();
        defer self.state_mutex.unlock();

        if (!self.connect_done) {
            self.connect_cond.timedWait(&self.state_mutex, timeout_ms * std.time.ns_per_ms) catch {};
        }

        if (!self.connected) {
            log.err("[Error] Connection timeout or refused ({s}:{d})", .{ self.cfg.host, self.cfg.port });
            self.running = false;
            _ = c.mosquitto_loop_stop(self.mosq, true);
            return Error.ConnectTimeout;
        }

        log.info("[Connected] Successfully connected to {s}:{d}", .{ self.cfg.host, self.cfg.port });
    }

    pub fn disconnect(self: *Self) void {
        self.running = false;
        _ = c.mosquitto_disconnect(self.mosq);
        _ = c.mosquitto_loop_stop(self.mosq, false);
        self.state_mutex.lock();
        defer self.state_mutex.unlock();
        self.connected = false;
    }

    pub fn deinit(self: *Self) void {
        if (self.running or self.connected) {
            self.disconnect();
        }
        c.mosquitto_destroy(self.mosq);
    }

    /// Write a single key/value pair. Returns true on success.
    pub fn writeSingle(
        self: *Self,
        topic: [:0]const u8,
        value: ?[]const u8,
        qos: Qos,
        retain: bool,
        timeout_ms: u64,
    ) bool {
        if (!self.connected) {
            log.err("[Error] Not connected. Call connect() first.", .{});
            return false;
        }

        const payload = value orelse "";
        var mid: c_int = 0;
        const rc = c.mosquitto_publish(
            self.mosq,
            &mid,
            topic.ptr,
            @intCast(payload.len),
            payload.ptr,
            @intFromEnum(qos),
            retain,
        );
        if (rc != c.MOSQ_ERR_SUCCESS) {
            log.err("[Error] Failed to queue message for {s}: {s}", .{ topic, c.mosquitto_strerror(rc) });
            return false;
        }

        const timeout_ns = timeout_ms * std.time.ns_per_ms;

        self.pending_mutex.lock();
        defer self.pending_mutex.unlock();

        if (self.pendingAdd(mid) == null) {
            log.err("[Error] Too many pending publishes", .{});
            return false;
        }

        const completed = self.pendingWait(mid, timeout_ns);
        var success = false;

        if (completed) {
            if (self.pendingFind(mid)) |idx| {
                success = self.pending[idx].success;
                self.pendingRemove(idx);
            }
        } else {
            log.err("[Timeout] Publish timeout for {s}", .{topic});
            if (self.pendingFind(mid)) |idx| {
                self.pendingRemove(idx);
            }
        }

        if (success) {
            log.info("[Written] {s} => {s}", .{ topic, payload });
        }

        return success;
    }

    /// Write multiple key/value pairs. Returns count of successes.
    pub fn writeBatch(
        self: *Self,
        topics: []const [:0]const u8,
        values: []const ?[]const u8,
        qos: Qos,
        retain: bool,
        timeout_ms: u64,
    ) usize {
        if (!self.connected) {
            log.err("[Error] Not connected. Call connect() first.", .{});
            return 0;
        }

        const count = @min(topics.len, MAX_BATCH);
        const timeout_ns = timeout_ms * std.time.ns_per_ms;

        var mids: [MAX_BATCH]c_int = [_]c_int{0} ** MAX_BATCH;
        var queued: [MAX_BATCH]bool = [_]bool{false} ** MAX_BATCH;

        // Phase 1: publish all messages and register mids
        self.pending_mutex.lock();
        for (0..count) |i| {
            const payload = values[i] orelse "";
            const rc = c.mosquitto_publish(
                self.mosq,
                &mids[i],
                topics[i].ptr,
                @intCast(payload.len),
                payload.ptr,
                @intFromEnum(qos),
                retain,
            );
            if (rc != c.MOSQ_ERR_SUCCESS) {
                log.err("[Error] Failed to queue {s}: {s}", .{ topics[i], c.mosquitto_strerror(rc) });
                continue;
            }
            if (self.pendingAdd(mids[i]) == null) {
                log.err("[Error] Too many pending publishes", .{});
                continue;
            }
            queued[i] = true;
        }
        self.pending_mutex.unlock();

        // Phase 2: wait for each queued message
        var success_count: usize = 0;

        self.pending_mutex.lock();
        defer self.pending_mutex.unlock();

        for (0..count) |i| {
            if (!queued[i]) continue;

            const completed = self.pendingWait(mids[i], timeout_ns);
            const idx = self.pendingFind(mids[i]);

            if (completed) {
                if (idx) |j| {
                    if (self.pending[j].success) {
                        success_count += 1;
                        log.info("[Batch Written] {s}", .{topics[i]});
                    }
                }
            } else {
                log.err("[Batch Timeout] {s}", .{topics[i]});
            }

            if (idx) |j| self.pendingRemove(j);
        }

        return success_count;
    }

    /// Delete a single key by publishing an empty retained message.
    pub fn deleteSingle(self: *Self, topic: [:0]const u8, timeout_ms: u64) bool {
        if (!self.connected) {
            log.err("[Error] Not connected. Call connect() first.", .{});
            return false;
        }

        var mid: c_int = 0;
        const rc = c.mosquitto_publish(self.mosq, &mid, topic.ptr, 0, "", 1, true);
        if (rc != c.MOSQ_ERR_SUCCESS) {
            log.err("[Error] Failed to delete {s}: {s}", .{ topic, c.mosquitto_strerror(rc) });
            return false;
        }

        const timeout_ns = timeout_ms * std.time.ns_per_ms;

        self.pending_mutex.lock();
        defer self.pending_mutex.unlock();

        if (self.pendingAdd(mid) == null) {
            log.err("[Error] Too many pending publishes", .{});
            return false;
        }

        const completed = self.pendingWait(mid, timeout_ns);
        var success = false;

        if (completed) {
            if (self.pendingFind(mid)) |idx| {
                success = self.pending[idx].success;
                self.pendingRemove(idx);
            }
        } else {
            log.err("[Timeout] Delete timeout for {s}", .{topic});
            if (self.pendingFind(mid)) |idx| {
                self.pendingRemove(idx);
            }
        }

        if (success) {
            log.info("[Deleted] {s}", .{topic});
        }

        return success;
    }

    /// Delete multiple keys. Returns count of successes.
    pub fn deleteBatch(
        self: *Self,
        topics: []const [:0]const u8,
        timeout_ms: u64,
    ) usize {
        if (!self.connected) {
            log.err("[Error] Not connected. Call connect() first.", .{});
            return 0;
        }

        const count = @min(topics.len, MAX_BATCH);
        const timeout_ns = timeout_ms * std.time.ns_per_ms;

        var mids: [MAX_BATCH]c_int = [_]c_int{0} ** MAX_BATCH;
        var queued: [MAX_BATCH]bool = [_]bool{false} ** MAX_BATCH;

        // Phase 1: publish all empty retained messages
        self.pending_mutex.lock();
        for (0..count) |i| {
            const rc = c.mosquitto_publish(self.mosq, &mids[i], topics[i].ptr, 0, "", 1, true);
            if (rc != c.MOSQ_ERR_SUCCESS) {
                log.err("[Error] Failed to queue delete for {s}: {s}", .{ topics[i], c.mosquitto_strerror(rc) });
                continue;
            }
            if (self.pendingAdd(mids[i]) == null) {
                log.err("[Error] Too many pending publishes", .{});
                continue;
            }
            queued[i] = true;
        }
        self.pending_mutex.unlock();

        // Phase 2: wait for each
        var success_count: usize = 0;

        self.pending_mutex.lock();
        defer self.pending_mutex.unlock();

        for (0..count) |i| {
            if (!queued[i]) continue;

            const completed = self.pendingWait(mids[i], timeout_ns);
            const idx = self.pendingFind(mids[i]);

            if (completed) {
                if (idx) |j| {
                    if (self.pending[j].success) {
                        success_count += 1;
                        log.info("[Batch Deleted] {s}", .{topics[i]});
                    }
                }
            } else {
                log.err("[Delete Timeout] {s}", .{topics[i]});
            }

            if (idx) |j| self.pendingRemove(j);
        }

        return success_count;
    }

    /// Update a key's value (convenience: writeSingle with retain=true).
    pub fn updateSingle(self: *Self, topic: [:0]const u8, value: []const u8, qos: Qos, timeout_ms: u64) bool {
        return self.writeSingle(topic, value, qos, true, timeout_ms);
    }

    pub fn isConnected(self: *Self) bool {
        self.state_mutex.lock();
        defer self.state_mutex.unlock();
        return self.connected;
    }

    // ── Mosquitto C callbacks ────────────────────────────────────────

    fn onConnectCb(_: ?*c.struct_mosquitto, userdata: ?*anyopaque, rc: c_int) callconv(.C) void {
        const self: *Self = @ptrCast(@alignCast(userdata));
        self.state_mutex.lock();
        defer self.state_mutex.unlock();
        self.connected = (rc == 0);
        self.connect_done = true;
        self.connect_cond.signal();
    }

    fn onDisconnectCb(_: ?*c.struct_mosquitto, userdata: ?*anyopaque, _: c_int) callconv(.C) void {
        const self: *Self = @ptrCast(@alignCast(userdata));
        self.state_mutex.lock();
        defer self.state_mutex.unlock();
        self.connected = false;
    }

    fn onPublishCb(_: ?*c.struct_mosquitto, userdata: ?*anyopaque, mid: c_int) callconv(.C) void {
        const self: *Self = @ptrCast(@alignCast(userdata));
        self.pending_mutex.lock();
        defer self.pending_mutex.unlock();
        if (self.pendingFind(mid)) |idx| {
            self.pending[idx].completed = true;
            self.pending[idx].success = true;
        }
        self.pending_cond.broadcast();
    }
};

// ──────────────────────────────────────────────────────────────────────
//  Tests
// ──────────────────────────────────────────────────────────────────────

test "Writer Config defaults" {
    const cfg = Config{};
    try std.testing.expectEqualStrings("localhost", cfg.host);
    try std.testing.expectEqual(@as(u16, 1883), cfg.port);
    try std.testing.expectEqualStrings("kv-writer", cfg.client_id);
    try std.testing.expect(cfg.clean_session);
}